readyState may not move from HAVE_METADATA following a seek.

jyavenard opened the pull request:
#### e1d76fd08756b9610c3a8340ee4f4eaabcc91707
<pre>
readyState may not move from HAVE_METADATA following a seek.
<a href="https://bugs.webkit.org/show_bug.cgi?id=254136">https://bugs.webkit.org/show_bug.cgi?id=254136</a>
rdar://problem/106914173

Reviewed by Youenn Fablet.

Seen with LayoutTests/media/media-source/media-source-end-of-stream-buffered.html
when running in the GPU process.
When we seeked to the end of the video while the MediaSource is still in
open state, the readyState moved to HAVE_METADATA.
Once MediaSource.endOfStream() was called the readyState should be moved
to HAVE_CURRENT_DATA (<a href="https://html.spec.whatwg.org/multipage/media.html#dom-media-have_current_data)">https://html.spec.whatwg.org/multipage/media.html#dom-media-have_current_data)</a>
However, we didn&apos;t update the readyState under those circumstances.
As such if play() was called, the waiting event would fire instead of play()

We partially have to revert webkit.org/b/253560 as in this change we
only re-calculated the buffered when queried.
We need to do so whenever the MediaSource&apos;s buffered attribute change
(such as when the MediaSource&apos;s readyState has changed) and for running
the SourceBuffer monitoring algorithm which will in turn update the
HTMLMediaElement&apos;s readyState accordingly.

Additionally, similar to bug 252886 (durationchanged to be fired befure updateend)
per spec <a href="https://w3c.github.io/media-source/#sourcebuffer-coded-frame-processing">https://w3c.github.io/media-source/#sourcebuffer-coded-frame-processing</a>
step 4, we have &quot;If the HTMLMediaElement.readyState attribute is
HAVE_CURRENT_DATA and the new coded frames cause HTMLMediaElement.buffered
to have a TimeRanges that includes the current playback position and some
time beyond the current playback position, then set the HTMLMediaElement
readyState attribute to HAVE_FUTURE_DATA.&quot;
This step is to happen before we end the coded frame processing algorithm
and we complete the appendBuffer() operation and fire `updateend` event.
As such, we make `sourceBufferPrivateBufferedChanged()` asynchronous so
that we resume the AppendComplete operation only after we have modified
the buffered range and processed the readyState change.
This change is required as the order of events changed with WK1 and to
keep test expectations we need to also update the behaviour for when the
GPU process is active.

We update some test expectations accordingly as events related to the readyState
changed will be fired before `updateend`, in particular events such as
`canplay` or `canplaythrough`.

LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-audio-only-expected.txt
LayoutTests/media/media-source/media-source-monitor-playing-event-expected.txt

* LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-audio-only-expected.txt:
* LayoutTests/media/media-source/media-source-allowed-codecs-expected.txt: mp4v is an invalid codec and reported as such in recent CoreMedia framework.
* LayoutTests/media/media-source/media-source-allowed-codecs.html:
* LayoutTests/media/media-source/media-source-monitor-playing-event-expected.txt:
* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::isBuffered const):
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::MediaSource):
(WebCore::MediaSource::buffered const):
(WebCore::MediaSource::onReadyStateChange):
(WebCore::MediaSource::sourceBufferBufferedChanged):
(WebCore::MediaSource::updateBufferedIfNeeded):
(WebCore::MediaSource::buffered): Deleted.
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::sourceBufferPrivateBufferedChanged):
(WebCore::SourceBuffer::setBufferedDirty):
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/platform/graphics/MediaSourcePrivate.h:
(WebCore::MediaSourcePrivate::bufferedChanged):
* Source/WebCore/platform/graphics/MediaSourcePrivateClient.h:
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::setBufferedRanges):
(WebCore::SourceBufferPrivate::appendCompleted):
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
(WebCore::SourceBufferPrivate::setBufferedRanges):
* Source/WebCore/platform/graphics/SourceBufferPrivateClient.h:
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:
(WebKit::RemoteMediaSourceProxy::buffered const):
(WebKit::RemoteMediaSourceProxy::buffered): Deleted.
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateBufferedChanged):
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateAppendComplete):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateAppendComplete):
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateBufferedChanged):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.messages.in:

Canonical link: <a href="https://commits.webkit.org/261949@main">https://commits.webkit.org/261949@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/254041d3769fdf6bd177e18f37d54888a7c9b06e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/113217 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22350 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1890 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4994 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121677 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/117350 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23720 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13528 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/6226 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/119004 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/17620 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100882 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/106309 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99574 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1423 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46667 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14649 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1463 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/98916 "9 api tests failed or timed out") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15361 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/10790 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20653 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53454 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/17211 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.h:
```diff
@@ -172,6 +172,7 @@ class MediaSource
     RefPtr<MediaSourcePrivate> m_private;
     RefPtr<SourceBufferList> m_sourceBuffers;
     RefPtr<SourceBufferList> m_activeSourceBuffers;
+    std::unique_ptr<PlatformTimeRanges> m_buffered;
```
Can m_buffered be null anymore?
If not, we can change it to a UniqueRef.

jyavenard replied:
ok

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -1147,15 +1145,27 @@ void MediaSource::notifyElementUpdateMediaState() const
     for (auto& sourceBuffer : *m_activeSourceBuffers)
         sourceBuffer->setBufferedDirty(false);
 
+    auto buffered = WTFMove(m_buffered);
```
Can buffered be a PlatformTimeRanges instead of a unique_ptr?

jyavenard replied:
> Can buffered be a PlatformTimeRanges instead of a unique_ptr?

That would be changing the API from SourceBufferPrivateClient ; I guess we could. But there are implementation that return nullptr right now.

jyavenard replied:
such as `std::unique_ptr<PlatformTimeRanges> RemoteMediaSourceProxy::buffered()` however, I am making changes in that code soon, following which it can never be null

jyavenard replied:
On second thoughts, MediaSource::buffered() changes a lot and often, you probably want to avoid anything taking a reference and not make a copy first. Particularly as there's re-entrancy everywhere in that code. You never know if an object won't get modified from under you.

youennf commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -91,13 +91,17 @@ void SourceBufferPrivate::updateHighestPresentationTimestamp()
-        m_client->sourceBufferPrivateBufferedChanged();
+        m_client->sourceBufferPrivateBufferedChanged(buffered(), WTFMove(completionHandler));
+    else
+        completionHandler();
```
I would write it:
```
if (!m_client) {
    completionHandler();
    return;
}
m_client...
```

youennf commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -156,29 +160,34 @@ void SourceBufferPrivate::appendCompleted(bool parsingSucceeded, bool isEnded, F
+            if (m_client) {
+                if (!m_didReceiveSampleErrored)
+                    m_client->sourceBufferPrivateAppendComplete(parsingSucceeded ? SourceBufferPrivateClient::AppendResult::Succeeded : SourceBufferPrivateClient::AppendResult::ParsingFailed);
+                m_client->sourceBufferPrivateReportExtraMemoryCost(totalTrackBufferSizeInBytes());
```
Are we sure this is still valid after the call to sourceBufferPrivateAppendComplete?
Should we protecThis weakSelf?

jyavenard committed:
readyState may not move from HAVE_METADATA following a seek.
https://bugs.webkit.org/show_bug.cgi?id=254136
rdar://problem/106914173

Reviewed by Youenn Fablet.

Seen with LayoutTests/media/media-source/media-source-end-of-stream-buffered.html
when running in the GPU process.
When we seeked to the end of the video while the MediaSource is still in
open state, the readyState moved to HAVE_METADATA.
Once MediaSource.endOfStream() was called the readyState should be moved
to HAVE_CURRENT_DATA (https://html.spec.whatwg.org/multipage/media.html#dom-media-have_current_data)
However, we didn't update the readyState under those circumstances.
As such if play() was called, the waiting event would fire instead of play()

We partially have to revert webkit.org/b/253560 as in this change we
only re-calculated the buffered when queried.
We need to do so whenever the MediaSource's buffered attribute change
(such as when the MediaSource's readyState has changed) and for running
the SourceBuffer monitoring algorithm which will in turn update the
HTMLMediaElement's readyState accordingly.

Additionally, similar to bug 252886 (durationchanged to be fired befure updateend)
per spec https://w3c.github.io/media-source/#sourcebuffer-coded-frame-processing
step 4, we have "If the HTMLMediaElement.readyState attribute is
HAVE_CURRENT_DATA and the new coded frames cause HTMLMediaElement.buffered
to have a TimeRanges that includes the current playback position and some
time beyond the current playback position, then set the HTMLMediaElement
readyState attribute to HAVE_FUTURE_DATA."
This step is to happen before we end the coded frame processing algorithm
and we complete the appendBuffer() operation and fire `updateend` event.
As such, we make `sourceBufferPrivateBufferedChanged()` asynchronous so
that we resume the AppendComplete operation only after we have modified
the buffered range and processed the readyState change.
This change is required as the order of events changed with WK1 and to
keep test expectations we need to also update the behaviour for when the
GPU process is active.

We update some test expectations accordingly as events related to the readyState
changed will be fired before `updateend`, in particular events such as
`canplay` or `canplaythrough`.

LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-audio-only-expected.txt
LayoutTests/media/media-source/media-source-monitor-playing-event-expected.txt

* LayoutTests/http/tests/media/fairplay/fps-mse-unmuxed-audio-only-expected.txt:
* LayoutTests/media/media-source/media-source-allowed-codecs-expected.txt: mp4v is an invalid codec and reported as such in recent CoreMedia framework.
* LayoutTests/media/media-source/media-source-allowed-codecs.html:
* LayoutTests/media/media-source/media-source-monitor-playing-event-expected.txt:
* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::isBuffered const):
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::MediaSource):
(WebCore::MediaSource::buffered const):
(WebCore::MediaSource::onReadyStateChange):
(WebCore::MediaSource::sourceBufferBufferedChanged):
(WebCore::MediaSource::updateBufferedIfNeeded):
(WebCore::MediaSource::buffered): Deleted.
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::sourceBufferPrivateBufferedChanged):
(WebCore::SourceBuffer::setBufferedDirty):
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/platform/graphics/MediaSourcePrivate.h:
(WebCore::MediaSourcePrivate::bufferedChanged):
* Source/WebCore/platform/graphics/MediaSourcePrivateClient.h:
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::setBufferedRanges):
(WebCore::SourceBufferPrivate::appendCompleted):
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
(WebCore::SourceBufferPrivate::setBufferedRanges):
* Source/WebCore/platform/graphics/SourceBufferPrivateClient.h:
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:
(WebKit::RemoteMediaSourceProxy::buffered const):
(WebKit::RemoteMediaSourceProxy::buffered): Deleted.
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateBufferedChanged):
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateAppendComplete):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateAppendComplete):
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateBufferedChanged):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.messages.in:

Canonical link: https://commits.webkit.org/261949@main
